XpressAI/xircuits
# ✨Enable branching component

# Description

This will enable branching workflow by adding `outFlow` port in any node. 

**How it works**
A branch component will have `outFlow` ports. This port will run **first** given condition(if any) its next node until there's no more node. (Let call it branch workflow)
The last node of this workflow can be left hanging or not connecting to any component/node. It doesn't need to end with Finish node(only for branch workflow). 
Then, it'll go back to the branch component to continue with its finished port (▶)(Finished workflow). This workflow **Must** end with `Finish` node.

**How to create a branch node**

1. Use any existing node or create new node.
2. Create a port name(e.g test-branch) as usual but instead of using `InArg` or `OutArg`, use **`BaseComponent`**. This will create an `outFlow` port(test-branch ▶).

> test_branch: BaseComponent

3. Declare in def __init__

> self.test_branch = BaseComponent

4. In def execute

> self.test_branch .do(ctx)

This will make it run the next component of test-branch ▶ port.

**Example of branches component**

There are two example of branches component:
    - TestFlowPort : Branching workflow with condition.
    - AlphabetSwitch : Branching workflow randomly.

**Minor details changes:**

- Indirectly bypass the `checkAllNodesConnected` when there's at least a connection from Start to Finish.
- Disable InPort accepting multiple links.
- Change the error tooltip UI due to blur bug.

## References

Remake of #159.

## Pull Request Type


## Type of Change


# Tests

1. Try creating an outFlow port.
    1. Follow as mentioned above.
2. Try creating multiple outFlow ports given condition.
    1. You can try creating your own or use the 2 example branch component.
    2. Try running xircuits using it. You can follow the description on how to use it.
3. (Additional) Xircuits can run when there's at least have one complete workflow(Start to Finish).
    1. Try creating a complete workflow.
    3. Add another component. Don't connect the node.
    4. Try running it. It's expected to work.
5. Try compiling workflow without ending with Finish node.
    1. An error tooltip will be display.
6. Try running without connecting one/both `outFlow` port but Finished port (▶) port must be connected.
    1. If either/both port not connected, it will run the Finished port (▶) port.
7. Try running without connecting Finished port (▶) port.
    1. An error tooltip will be display.
8. Try different and crazy but feasible workflow. (Let me know if it doesn't work)

## Tested on?


# Notes

Currently, it's not possible to pass variable at the end of branch workflow to the finish workflow
